### PR TITLE
Run tests on snippets instead of files

### PR DESCRIPTION
**Pull request by fisker**
Some codes can't test pieces together, for example, we can't test two json string in one file.

```js
run_spec(
  {
    dirname: __dirname,
    codes: [
      {
        name: "JSON object",
        code: '{"foo": 1}',
      },
      {
        name: "JSON array",
        code: '["foo", "bar"]',
      },
    ],
  },
  ["json"]
);
```

Some codes testing special characters, we can use js to create those files.

```js
run_spec(
  {
    dirname: __dirname,
    codes: [
      {
        name: "empty with BOM",
        code: "\uFEFF",
      },
      {
        name: "BOM",
        code: `\uFEFF
          const foo = bar
        `,
      },
      {
        name: "`CRLF` as linebreak",
        code: "const foo=1;\r\nconst bar=2;",
      },
    ],
  },
  ["babel"]
);
```

We don't need create small files anymore, or worry about editor replacing those special characters.

`name` and `filename` both optional

```js
run_spec(
  {
    dirname: __dirname,
    codes: [
      // just code
      'code snippet 1',
      'code snippet 2',
    ]
  },
  ["babel"]
)
```

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: tests/js_empty/jsfmt.spec.js [hunk 1]
```diff
@@ -1 +1,18 @@
...
+      // empty
+      "",
+      // empty lines
+      "\n",
+      "\n\n\n\n",
+      // semicolons
+      ";",
+      ";;;;",
+      ";\n",
+      ";\n\n;;;\n",
```
Comment (thorn0) [inline tests/js_empty/jsfmt.spec.js:14]: This looks fine.

File: tests/parser-inference/jsfmt.spec.js [hunk 2]
```diff
@@ -1 +1,48 @@
...
+  dirname: __dirname,
+  codes: [
+    {
+      name: "html",
+      filename: "prettier.html",
+      code: dedent`
+        <foo>
+                        <bar/>
+        </foo>
+      `,
```
Comment (thorn0) [inline tests/parser-inference/jsfmt.spec.js:13]: But this looks worse than files IMHO. There is nothing wrong with small files. They should be small.

**Comment by thorn0** [comment]
Also let's use `code` or `snippets` instead of `codes`. The word code is uncountable when it means "source code".

File: tests/parser-inference/jsfmt.spec.js (hunk 2 above)
Comment (fisker) [reply tests/parser-inference/jsfmt.spec.js]: I restored them.

**Comment by fisker** [comment]
Sorry for my poor english, 

```js
{
  code: [
    {
      code: ''
    }
  ]
}
```

seems weird, so `snippets`?


[Eslint called their tests `scenarios`](https://github.com/eslint/eslint/blob/b398e0b9ff455e4e7d70f19d5ccea5819c21eb86/lib/rule-tester/rule-tester.js#L436) , I don't know if it's a good name?

**Comment by thorn0** [comment]
`snippets` is fine.

**Comment by fisker** [comment]
@thorn0 I checked the `index.js` file in `dist`,  on L24254

```js
let {
  resolve: resolve$1
} = eval("require");
```

I remember this can't run in atom?

File: tests_config/run_spec.js [hunk 3]
```diff
@@ -67,32 +70,50 @@ global.run_spec = (dirname, parsers, options) => {
...
-      !file.isFile() ||
-      basename[0] === "." ||
-      basename === "jsfmt.spec.js"
-    ) {
-      continue;
-    }
+  const snippets = (fixtures.snippets || []).map((test, index) => {
+    test = typeof test === "string" ? { code: test } : test;
+    return {
+      name: `code: ${test.name || `#${index}`}`,
```
Comment (thorn0) [inline tests_config/run_spec.js:76]: Would be logical:
```suggestion
      name: `snippet: ${test.name || `#${index}`}`,
```

**Comment by fisker** [comment]
I need add this to https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md

**Pull request merged**